src/Microsoft.OData.Edm/Csdl/Parsing/CsdlJsonParser.cs

```diff
@@ -0,0 +1,335 @@
+                {
+                    switch (propertyName)
+                    {
+                        case "$Version":
+                            // skip, because processed
+                            break;
+
+                        case "$EntityContainer":
+                            // The value of $EntityContainer is the namespace-qualified name of the entity container of that service.
+                            // So far, i don't know how to use it. So skip it.
+                            break;
+
+                        case "$Reference":
+                            // The document object MAY contain the member $Reference to reference other CSDL documents.
+                            references = ParseReferences(propertyValue, context);
+                            break;
+
+                        default:
+                            // CSDL document also MAY contain members for schemas.
+                            // Each schema's value is an object.
+                            if (propertyValue.ValueKind == JsonValueKind.Object)
+                            {
+                                CsdlSchema schema = SchemaJsonParser.ParseCsdlSchema(propertyName, csdlModel.CsdlVersion, propertyValue, context);
+                                if (schema != null)
+                                {
+                                    csdlModel.AddSchema(schema);
+                                    break;
+                                }
+                            }
+
+                            context.ReportError(EdmErrorCode.JsonUnexpectedElement, Strings.CsdlJsonParser_UnexpectedJsonMember(context.Path, propertyValue.ValueKind));
+                            break;
+                    }
+                });
+
+
+                if (references != null)
+                {
+                    csdlModel.AddCurrentModelReferences(references);
+                }
+
+                return csdlModel;
+            }
+        }
+
+        /// <summary>
+        /// Build the JSON value <see cref="JsonElement"/> to collection of <see cref="IEdmReference"/>.
+        /// It's weird that we parse it into IEdmReference, not CsdlReference.
+        /// It is because CsdlModel needs "IEdmReference". :-(
+        /// </summary>
+        /// <param name="element">The input JSON element.</param>
+        /// <param name="jsonPath">The parser context.</param>
+        /// <returns>null or parsed collection of <see cref="IEdmReference"/>.</returns>
+        internal static IList<IEdmReference> ParseReferences(JsonElement element, JsonParserContext context)
+        {
+            // The value of $Reference is an object that contains one member per referenced CSDL document.
+            if (!element.ValidateValueKind(JsonValueKind.Object, context))
+            {
+                return null;
+            }
+
+            IList<IEdmReference> references = new List<IEdmReference>();

```

Comment:
>IList [](start = 12, length = 5)

Why IList and not List here?

Reply:
Most of my usage is to use IList<T> to refer an List<T> instance.
Except is taht if i want to call "AddRange", i have to use List<T> to refer to a List<T> instance.

---
In reply to: [443022342](https://github.com/OData/odata.net/pull/1775#discussion_r443022342) [](ancestors = 443022342)